src/main/java/ch/njol/skript/expressions/ExprSignText.java

```diff
@@ -32,145 +26,203 @@
-		return String.class;
-	}
-	
+
 	@Override
 	@Nullable
-	protected String[] get(final Event e) {
-		final Number l = line.getSingle(e);
-		if (l == null)
-			return new String[0];
-		final int line = l.intValue() - 1;
-		if (line < 0 || line > 3)
-			return new String[0];
-		if (getTime() >= 0 && block.isDefault() && e instanceof SignChangeEvent && !Delay.isDelayed(e)) {
-			return new String[] {((SignChangeEvent) e).getLine(line)};
+	protected String[] get(Event event) {
+		Block block = exprBlock.getSingle(event);
+		if (isLines) {
+			if (block == null || !sign.isOfType(block))
+				return new String[0];
+			if ((event instanceof SignChangeEvent && getTime() >= 0 && block.equals(((SignChangeEvent) event).getBlock()) && !delayed))
+				return ((SignChangeEvent) event).getLines();
+			return ((Sign) (block.getState())).getLines();
+		} else {
+			Number lineNumber = line.getSingle(event);
+			if (lineNumber == null || lineNumber.intValue() < 1 || lineNumber.intValue() > 4)
+				return new String[0];
+			int index = lineNumber.intValue() - 1;
+			if (event instanceof SignChangeEvent && getTime() >= 0 && block == ((SignChangeEvent) event).getBlock() && !delayed) {
+				return CollectionUtils.array(((SignChangeEvent) event).getLine(index));
+			}
+			if (block == null || !sign.isOfType(block)) {
+				return new String[0];
+			}
+			return CollectionUtils.array(
+				((Sign) (block.getState())).getLine(index)
+			);
 		}
-		final Block b = block.getSingle(e);
-		if (b == null)
-			return new String[0];
-		if (!sign.isOfType(b))
-			return new String[0];
-		return new String[] {((Sign) b.getState()).getLine(line)};
 	}
-	
-	@Override
-	public String toString(final @Nullable Event e, final boolean debug) {
-		return "line " + line.toString(e, debug) + " of " + block.toString(e, debug);
-	}
-	
-	// TODO allow add, remove, and remove all (see ExprLore)
+
 	@Override
 	@Nullable
 	public Class<?>[] acceptChange(final ChangeMode mode) {
-		if (mode == ChangeMode.DELETE || mode == ChangeMode.SET)
-			return new Class[] {String.class};
+		if (isLines) {
+			if (mode == ChangeMode.RESET) {
+				return null;
+			}
+			return CollectionUtils.array(String[].class);
+		} else if (mode == ChangeMode.SET || mode == ChangeMode.DELETE) {
+			return CollectionUtils.array(String.class);
+		}
 		return null;
 	}
-	
-	static boolean hasUpdateBooleanBoolean = true;
-	
-	@SuppressWarnings("incomplete-switch")
+
 	@Override
-	public void change(final Event e, final @Nullable Object[] delta, final ChangeMode mode) throws UnsupportedOperationException {
-		final Number l = line.getSingle(e);
-		if (l == null)
-			return;
-		final int line = l.intValue() - 1;
-		if (line < 0 || line > 3)
-			return;
-		final Block b = block.getSingle(e);
-		if (b == null)
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
+		if (delta == null) {
+			if (mode != ChangeMode.DELETE && mode != ChangeMode.RESET)
+				return;
+		}
```

Comment:
Not needed, Skript will do this entire condition for you.

Reply:
Null checks on delta is not needed anymore? As in my whitelist PR [I was told](https://github.com/SkriptLang/Skript/pull/5422/files#r1100590025) to keep the null check